### ISSUE TITLE
sudo systemctl stop k3s not worked as expected

### ISSUE THREAD

**Issue by dfang** [k3s-io/k3s#2583]
**Environmental Info:**
K3s Version: 

```
$ k3s -v
k3s version v1.19.3+k3s3 (0e4fbfef)
```

Node(s) CPU architecture, OS, and Version: 

```
$ uname -a
Linux ip-172-31-43-180.ap-east-1.compute.internal 4.14.203-156.332.amzn2.x86_64 #1 SMP Fri Oct 30 19:19:33 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

latest Linux AMI 2 on aws

Cluster Configuration:

how i install k3s

```
curl -sfL https://get.k3s.io | sh -
sudo chmod 644 /etc/rancher/k3s/k3s.yaml
```

**Describe the bug:**

stop service not work as expected

```
$ sudo systemctl  stop k3s
$ sudo systemctl status k3s
● k3s.service - Lightweight Kubernetes
   Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Tue 2020-11-24 03:19:45 UTC; 5min ago
     Docs: https://k3s.io
  Process: 5654 ExecStart=/usr/local/bin/k3s server (code=exited, status=1/FAILURE)
  Process: 5652 ExecStartPre=/sbin/modprobe overlay (code=exited, status=0/SUCCESS)
  Process: 5650 ExecStartPre=/sbin/modprobe br_netfilter (code=exited, status=0/SUCCESS)
 Main PID: 5654 (code=exited, status=1/FAILURE)
    Tasks: 58
   Memory: 87.6M
   CGroup: /system.slice/k3s.service
           ├─3427 /var/lib/rancher/k3s/data/223e6420f8db0d8828a8f5ed3c44489bb8eb47aa71485404f8af8c462a29bea3/bin/containerd-shim-runc-v2 -namespace k8s.io ...
           ├─3570 /var/lib/rancher/k3s/data/223e6420f8db0d8828a8f5ed3c44489bb8eb47aa71485404f8af8c462a29bea3/bin/containerd-shim-runc-v2 -namespace k8s.io ...
           ├─3628 /var/lib/rancher/k3s/data/223e6420f8db0d8828a8f5ed3c44489bb8eb47aa71485404f8af8c462a29bea3/bin/containerd-shim-runc-v2 -namespace k8s.io ...
           ├─4411 /var/lib/rancher/k3s/data/223e6420f8db0d8828a8f5ed3c44489bb8eb47aa71485404f8af8c462a29bea3/bin/containerd-shim-runc-v2 -namespace k8s.io ...
           └─4613 /var/lib/rancher/k3s/data/223e6420f8db0d8828a8f5ed3c44489bb8eb47aa71485404f8af8c462a29bea3/bin/containerd-shim-runc-v2 -namespace k8s.io ...

Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: I1124 03:19:45.079639    5654 network_policy_controller.go:173] Shutting down ...roller
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: time="2020-11-24T03:19:45.079913551Z" level=error msg="Tunnel endpoint watch c...ontext
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: time="2020-11-24T03:19:45.080015821Z" level=info msg="Shutting down batch/v1, ...rkers"
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: time="2020-11-24T03:19:45.080080987Z" level=info msg="Shutting down helm.cattl...rkers"
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: time="2020-11-24T03:19:45.080109974Z" level=info msg="Shutting down helm.cattl...rkers"
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal k3s[5654]: time="2020-11-24T03:19:45.080135541Z" level=fatal msg="controllers exited"
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal systemd[1]: k3s.service: main process exited, code=exited, status=1/FAILURE
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal systemd[1]: Stopped Lightweight Kubernetes.
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal systemd[1]: Unit k3s.service entered failed state.
Nov 24 03:19:45 ip-172-31-43-180.ap-east-1.compute.internal systemd[1]: k3s.service failed.
Hint: Some lines were ellipsized, use -l to show in full.
```

**Steps To Reproduce:**

- Installed K3s: 

**Expected behavior:**

**Actual behavior:**
stop service not worked as expected

**Comment by brandond** [comment]
Stopping k3s does not stop workload containers. This is to allow for non disruptive upgrades. Can you clarify what you were expecting to see when you stopped the service?

**Comment by dfang** [comment]
after `sudo systemctl stop k3s`,  the status is failed ?

**Comment by brandond** [comment]
If it failed to start properly stopping it will not clear that status. Can you provide full logs as requested in your other issue?

**Comment by janeczku** [comment]
Reproduced on v1.19.3 with following unit file:

```
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
Wants=network-online.target
After=network-online.target
Wants=docker.service
After=docker.service

[Install]
WantedBy=multi-user.target

[Service]
Type=exec
Delegate=yes
RestartSec=5s
ExecStart=/usr/local/bin/k3s \
    server \
    "--docker" \
```

Before stopping status is active:

```
[~]# systemctl status k3s
  k3s.service - Lightweight Kubernetes
   Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: disabled)
   Active: active (running) since Tue 2020-11-24 17:16:38 UTC; 2min 29s ago
     Docs: https://k3s.io
 Main PID: 5967 (k3s-server)
    Tasks: 0
   Memory: 277.5M
   CGroup: /system.slice/k3s.service
           ‣ 5967 /usr/local/bin/k3s server
```

After stopping the service status is failed:

```
[~]# systemctl stop k3s
[~]# systemctl status k3s
  k3s.service - Lightweight Kubernetes
   Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Tue 2020-11-24 17:19:11 UTC; 1s ago
     Docs: https://k3s.io
  Process: 5967 ExecStart=/usr/local/bin/k3s server --docker (code=exited, status=1/FAILURE)
 Main PID: 5967 (code=exited, status=1/FAILURE)

Nov 24 17:19:11 test-centos-k3s-docker k3s[5967]: time="2020-11-24T17:19:11.651544109Z" level=fatal msg="controllers exited"
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: Child 5967 belongs to k3s.service
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: Unwatching 5967.
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: k3s.service: main process exited, code=exited, status=1/FAILURE
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: k3s.service changed stop-sigterm -> failed
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: Job k3s.service/stop finished, result=done
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: Stopped Lightweight Kubernetes.
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: Unit k3s.service entered failed state.
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: k3s.service failed.
Nov 24 17:19:11 test-centos-k3s-docker systemd[1]: k3s.service: cgroup is empty
```

**Comment by brandond** [comment]
That's probably because k3s shuts down with a nonzero exit code:
`Nov 24 17:19:11 test-centos-k3s-docker k3s[5967]: time="2020-11-24T17:19:11.651544109Z" level=fatal msg="controllers exited"`

Is this causing any problems for you, or is it merely a cosmetic issue?

**Comment by brandond** [comment]
Yes, the failed states is probably related to the cgroup issue in #2548

**Comment by ohdearaugustin** [comment]
I don't think that a non-zero status is only a cosmetic issue. You are not able to run k3s probably with systemd then.

I tried an update to 1.19.5 k3s1. The problem still remain.

I'm using following systemd unit:

```
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
After=network-online.target

[Service]
Type=notify
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
EnvironmentFile=/etc/default/k3s
ExecStart=/usr/local/bin/k3s server -i 10.10.20.1 --bind-address 10.10.20.1 --https-listen-port 6443 --flannel-iface ens10 --advertise-address 10.10.20.1 --kube-scheduler 'address=127.0.0.1' --kube-controller-manager-arg 'address=127.0.0.1' --kubelet-arg 'address=127.0.0.1' --cluster-init --disable traefik --disable servicelb  --disable local-storage
KillMode=process
ExecStopPost=/usr/local/bin/k3s-stop
#SuccessExitStatus=0 1
Delegate=yes
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ProtectHome=true

[Install]
WantedBy=multi-user.target
```

There are two scenarios. The first one when I stop k3s in the starting process, will end with an exit code 1. 
```Dec 15 00:22:50 <host_name: k3s[2290347]: time="2020-12-15T00:22:50.967424632+01:00" level=error msg="Tunnel context canceled while waiting for connection"
Dec 15 00:22:50 <host_name> k3s[2290347]: time="2020-12-15T00:22:50.967956516+01:00" level=fatal msg="context canceled"
Dec 15 00:22:51 <host_name> systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
```

Second a fully started k3s:

```
Dec 15 00:27:16 <host_name> k3s[2290609]: time="2020-12-15T00:27:16.831801983+01:00" level=fatal msg="leaderelection lost for k3s"
Dec 15 00:27:16 <host_name> systemd[1]: Stopping Lightweight Kubernetes...
Dec 15 00:27:16 <host_name> systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
```

I would rather see it as a problem of a gracefully shutdown. I couldn't relate my problems to #2548.
```

**Comment by brandond** [comment]
use k3s-killall.sh if you want to stop all the containers. It is intentional that stopping K3s and containerd does not stop workload pods.

**Comment by rafis** [comment]
I have found that there is CLI for containerd called `ctr` :-) I can manage containerd containers with it. It is enough for me:
```
ctr containers ls
ctr containers kill ...
ctr containers delete ...
```

**Comment by maa-biolinq** [comment]
I just started using **k3s** inside a Vagrant VM running Ubuntu 20.10, and I have been running into a similar issue: after installing k3s, the VM no longer shuts down gracefully -- vagrant times out and force-reboots the VM.

After reading through this thread, I solved my issue by adding a oneshot shutdown service that runs `k3s-killall.sh` when the system is trying to shutdown and allows for a graceful shutdown. `systemctl stop / reload k3s` behavior is unchanged and keeps the workloads running.

From my provisioning script:
```sh
cat >>/etc/systemd/system/k3s-shutdown.service <<END
[Unit]
Description=Gracefully shutdown k3s workload
DefaultDependencies=no
Before=shutdown.target

[Service]
Type=oneshot
ExecStart=/usr/local/bin/k3s-killall.sh
TimeoutStartSec=0

[Install]
WantedBy=shutdown.target
END

systemctl daemon-reload
systemctl enable k3s-shutdown.service
```

Maybe that could be added to the k3s setup script if deemed an acceptable solution ...
